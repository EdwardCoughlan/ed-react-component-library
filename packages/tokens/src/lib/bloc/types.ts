/* eslint-disable @typescript-eslint/no-explicit-any */
import { FunctionComponent, ComponentClass } from 'react';

export type ComponentMapper = {
  component: FunctionComponent<any> | ComponentClass<any>;
  mapper?: (props: any) => any;
};

export type ComponentMapperType = {
  [key: string]: ComponentMapper;
};
