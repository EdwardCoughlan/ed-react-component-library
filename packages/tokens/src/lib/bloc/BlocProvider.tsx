import {
  createContext,
  FC, ReactNode, useMemo,
  useRef,
} from 'react';
import { ComponentMapper, ComponentMapperType } from './types';

type BlocContextValue = {
  componentMappers: ComponentMapperType;
  addOrOverwriteComponent:(
    key: string,
    componentToAdd: ComponentMapper,
  ) => void;
};

export const BlocContext = createContext<BlocContextValue>({
} as BlocContextValue);

export const BlocProvider: FC<{
  componentMappers: ComponentMapperType;
  children: ReactNode;
}> = ({ componentMappers, children }) => {
  const componentMappersRef = useRef<ComponentMapperType>(componentMappers);

  const addOrOverwriteComponent = (
    key: string,
    componentToAdd: ComponentMapper,
  ): void => {
    if (componentMappersRef && componentMappersRef.current) {
      componentMappersRef.current[key] = componentToAdd;
    }
  };

  const value = useMemo(() => ({
    componentMappers: componentMappersRef.current,
    addOrOverwriteComponent,
  }), []);

  return (
    <BlocContext.Provider
      value={value}
    >
      {children}
    </BlocContext.Provider>
  );
};

export default BlocProvider;
