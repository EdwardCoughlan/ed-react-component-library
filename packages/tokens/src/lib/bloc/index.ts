export * from './bloc';
export * from './bloc-provider';
export * from './types';
