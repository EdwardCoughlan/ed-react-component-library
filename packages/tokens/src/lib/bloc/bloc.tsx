import {
  createElement,
  useContext,
} from 'react';
import { BlocContext } from './BlocProvider';
import { ComponentMapperType } from './types';

export type BlocProps = {
  id: string;
  props?: unknown;
  componentMapperIdentifier: string;
  componentMappers?: ComponentMapperType;
};

export const BlocFn = ({
  id,
  props,
  componentMapperIdentifier,
  componentMappers = {},
}: BlocProps): JSX.Element => {
  if (typeof componentMappers[componentMapperIdentifier] === 'undefined') {
    return createElement(
      () => {
        // eslint-disable-next-line no-console
        console.error(
          `Component not added to component mapper ${componentMapperIdentifier}. Try adding it or creating it.`,
        );
        return null;
      },
      { key: id },
    );
  }
  const componentMapper = componentMappers[componentMapperIdentifier];
  const componentProps = !!props && !!componentMapper.mapper
    ? componentMapper.mapper(props)
    : props;
  componentMapper.component.displayName = `${componentMapperIdentifier
    .charAt(0)
    .toUpperCase()}${componentMapperIdentifier.slice(1)}`;
  const retval = createElement(componentMapper.component, {
    ...componentProps,
    key: id,
  });
  return retval;
};

export function Bloc({
  componentMappers,
  ...restOfProps
}: BlocProps): JSX.Element {
  const context = useContext(BlocContext);
  if (context && context.componentMappers) {
    const contextLevelComponentMappers = context.componentMappers;
    if (componentMappers) {
      Object.keys(componentMappers).forEach((key) => {
        contextLevelComponentMappers[key] = componentMappers[key];
      });
    }
    return BlocFn({
      componentMappers: contextLevelComponentMappers,
      ...restOfProps,
    });
  }
  return BlocFn({ componentMappers, ...restOfProps });
}
