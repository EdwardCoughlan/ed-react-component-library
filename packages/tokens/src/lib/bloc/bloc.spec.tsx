import React, { FC, useContext } from 'react';
import { render } from '@testing-library/react';
import { Bloc } from './bloc';
import { BlocContext, BlocProvider } from './bloc-provider';
import { ComponentMapper } from './types';

const ComponentWithNoProps: React.FC = () => <div>Test with no props</div>;

type ComponentProps = {
  contents: string;
};
type ComponentProps2 = {
  alteredContents: string;
};
type ComponentProps2Altered = {
  alteredContents: string;
};

const ComponentWithProps = ({ contents }: ComponentProps): JSX.Element => (
  <div>{contents}</div>
);
const ComponentWithProps2: React.FC<ComponentProps2> = ({
  alteredContents,
}: ComponentProps2Altered) => <div>{alteredContents}</div>;
const ComponentWithProps3: React.FC<ComponentProps2> = ({
  alteredContents,
}: ComponentProps2Altered) => <div>{alteredContents}</div>;

const Mappers: Record<string, (props: ComponentProps) => ComponentProps2> = {
  componentWithProps2: (props): ComponentProps2 => ({
    alteredContents: `Altered ${(props as ComponentProps).contents}`,
  }),
  componentWithProps3: () => {
    throw new Error('Oh goodness gracious');
  },
};

const componentMappers: Record<string, ComponentMapper> = {
  componentWithNoProps: { component: ComponentWithNoProps },
  componentWithProps: { component: ComponentWithProps },
  componentWithProps2: {
    component: ComponentWithProps2,
    mapper: Mappers.componentWithProps2,
  },
  componentWithProps3: {
    component: ComponentWithProps3,
    mapper: Mappers.componentWithProps3,
  },
};

test('BLoc component without props', () => {
  const { container } = render(
    <Bloc
      id="1"
      componentMapperIdentifier="componentWithNoProps"
      componentMappers={componentMappers}
    />
  );

  expect(container).toMatchSnapshot();
});

test('Bloc component with props', () => {
  const test = "I am a component with props. That doesn't use a mapper";
  const { container } = render(
    <Bloc
      id="1"
      componentMapperIdentifier="componentWithProps"
      props={{
        contents: test,
      }}
      componentMappers={componentMappers}
    />
  );
  expect(container).toMatchSnapshot();
});

test("BLoc component doesn't exists", () => {
  const logSpy = jest.spyOn(console, 'error');
  render(
    <Bloc
      id="1"
      props={{
        content: "I am a component that doesn't exists",
      }}
      componentMapperIdentifier="componentThatDoesNotExists"
      componentMappers={componentMappers}
    />
  );
  expect(logSpy).toHaveBeenCalled();
});

test('Bloc component with props to be mapped', () => {
  const { container } = render(
    <Bloc
      id="1"
      componentMapperIdentifier="componentWithProps2"
      props={{
        contents: 'I am a component with props. That that will be mapped',
      }}
      componentMappers={componentMappers}
    />
  );
  expect(container).toMatchSnapshot();
});

test('Bloc component with props to be mapped error', () => {
  expect(() =>
    render(
      <Bloc
        id="1"
        componentMapperIdentifier="componentWithProps3"
        props={{
          contents:
            'I am a component with props. That that will be mapped but an error will happen',
        }}
        componentMappers={componentMappers}
      />
    )
  ).toThrowError();
});

test('Bloc with component coming from BlocContext', () => {
  type OldProps = { contents: string };
  type NewProps = { fromContext: string };
  const ContextComponent: React.FC<NewProps> = ({
    fromContext,
  }: {
    fromContext: string;
  }) => <div>{`context : ${fromContext}`}</div>;
  const { container } = render(
    <BlocProvider
      componentMappers={{
        componentFromContext: {
          component: ContextComponent,
          mapper: ({ contents }: OldProps): NewProps => ({
            fromContext: contents,
          }),
        },
      }}
    >
      <Bloc
        id="1"
        componentMapperIdentifier="componentFromContext"
        props={{
          contents: 'I am a component that is a part of the global context.',
        }}
        componentMappers={componentMappers}
      />
    </BlocProvider>
  );
  expect(container).toMatchSnapshot();
});

test('Bloc with BlocContext being overridden by inner value', () => {
  type OldProps = { contents: string };
  type NewProps = { fromContext: string };
  const ContextComponent: React.FC<NewProps> = ({
    fromContext,
  }: {
    fromContext: string;
  }) => <div>{`context : ${fromContext}`}</div>;
  const { container } = render(
    <BlocProvider
      componentMappers={{
        sameIdentifier: {
          component: ContextComponent,
          mapper: ({ contents }: OldProps): NewProps => ({
            fromContext: contents,
          }),
        },
      }}
    >
      <Bloc
        id="1"
        componentMapperIdentifier="sameIdentifier"
        props={{
          contents: 'I should have a prefix of override from bloc',
        }}
        componentMappers={{
          sameIdentifier: {
            component: ({ contents }: { contents: string }): JSX.Element => (
              <div>{`override from block: ${contents}`}</div>
            ),
          },
        }}
      />
    </BlocProvider>
  );
  expect(container).toMatchSnapshot();
});

test('No component mappers secified', () => {
  const logSpy = jest.spyOn(console, 'error');
  render(
    <Bloc
      id="1"
      componentMapperIdentifier="componentFromContext"
      props={{
        contents: 'I am a component that is a part of the global context.',
      }}
    />
  );

  expect(logSpy).toHaveBeenCalled();
});

test('complex prodvider setup', () => {
  const ComponentOne: FC = () => <div>I am Component One</div>;
  const BlocComponent: FC = () => (
    <div>
      <p>I am Bloc Component</p>
      <Bloc componentMapperIdentifier="ComponentOne" id="component-one" />
    </div>
  );
  const BlocComponentToAddComponents: FC = () => {
    const { addOrOverwriteComponent } = useContext(BlocContext);
    addOrOverwriteComponent('BlocComponent', { component: BlocComponent });
    return (
      <div>
        <p>I am the component that add another component to the context</p>
        <Bloc componentMapperIdentifier="BlocComponent" id="bloc-component" />
      </div>
    );
  };

  const { container } = render(
    <BlocProvider
      componentMappers={{
        BlocComponentToAddComponents: {
          component: BlocComponentToAddComponents,
        },
        ComponentOne: {
          component: ComponentOne,
        },
      }}
    >
      <Bloc
        componentMapperIdentifier="BlocComponentToAddComponents"
        id="bloc-component-to-add-components"
      />
    </BlocProvider>
  );
  expect(container).toMatchSnapshot();
});
