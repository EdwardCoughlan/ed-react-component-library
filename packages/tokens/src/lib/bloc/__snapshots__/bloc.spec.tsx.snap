// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BLoc component without props 1`] = `
<div>
  <div>
    Test with no props
  </div>
</div>
`;

exports[`Bloc component with props 1`] = `
<div>
  <div>
    I am a component with props. That doesn't use a mapper
  </div>
</div>
`;

exports[`Bloc component with props to be mapped 1`] = `
<div>
  <div>
    Altered I am a component with props. That that will be mapped
  </div>
</div>
`;

exports[`Bloc with BlocContext being overridden by inner value 1`] = `
<div>
  <div>
    override from block: I should have a prefix of override from bloc
  </div>
</div>
`;

exports[`Bloc with component coming from BlocContext 1`] = `
<div>
  <div>
    context : I am a component that is a part of the global context.
  </div>
</div>
`;

exports[`complex prodvider setup 1`] = `
<div>
  <div>
    <p>
      I am the component that add another component to the context
    </p>
    <div>
      <p>
        I am Bloc Component
      </p>
      <div>
        I am Component One
      </div>
    </div>
  </div>
</div>
`;
